### ISSUE TITLE
New crate 1.16.0 has panic in take.rs

### ISSUE THREAD

**Issue by kitsuneninetails** [tokio-rs/tokio#4435]
**Version**
1.16.0

**Platform**
Linux (local machine and in CircleCI)

**Description**
Panic in take under certain circumstances (using rocket):
```
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `0x56193d9ed894`,
 right: `0x56193d9ed890`', /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.0/src/io/util/take.rs:93:9
```

Apologies, we are using "rocket" webserver, so this isn't a minimal test case.

**Comment by kitsuneninetails** [comment]
I don't know the rocket usage under the covers that might isolate the specific usage of this take function.

**Comment by LEXUGE** [comment]
I can reproduce here.

**Comment by Darksonn** [comment]
Sorry about that. It will be fixed by #4437, which I will release as soon as possible.
